### PR TITLE
blog: Add GSoC'24 blog post on porting the mimalloc memory allocator to Unikraft

### PR DESCRIPTION
**Pull request by huyang531**
Signed-off-by: Yang Hu <yanghuu531@gmail.com>

### CODE REVIEW

File: content/blog/2024-06-16-unikraft-gsoc-port-mimalloc-to-unikraft.mdx [hunk 1]
```diff
@@ -0,0 +1,63 @@
+---
+title: "GSoC'24: Porting the mimalloc memory allocator to Unikraft"
+description: |
+  Different applications have different memroy usage patters and work best
```
Comment (RaduNichita) [inline content/blog/2024-06-16-unikraft-gsoc-port-mimalloc-to-unikraft.mdx:4]:
```suggestion
  Different applications have different memory usage patters and perform better
```

File: content/blog/2024-06-16-unikraft-gsoc-port-mimalloc-to-unikraft.mdx [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  using dedicated memory allocators. Right now, the Buddy allocator is the only
```
Comment (RaduNichita) [inline content/blog/2024-06-16-unikraft-gsoc-port-mimalloc-to-unikraft.mdx:5]:
```suggestion
  when using dedicated memory allocators. Right now, the buddy allocator is the only
```

File: content/blog/2024-06-16-unikraft-gsoc-port-mimalloc-to-unikraft.mdx [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+# Project Overview
+
+[mimalloc](https://github.com/microsoft/mimalloc) (pronounced "me-malloc") is a high-performance general-purpose memory allocator developed by Microsoft. [Hugo Lefeuvre](https://github.com/hlef) has made an effort to port mimalloc to Unikraft back in 2020 (see [this repo](https://github.com/unikraft/lib-mimalloc)). However, as Unikraft has evolved significantly over the years, more work is needed to adapt the `lib-mimalloc` port to the latest Unikraft core.
+
+## Objectives
+
+The steps of porting the memory allocators are as follows:
+
+1. Adapt the existing port, which uses `lib-newlib` and `lib-pthread-embedded` as the C library to use `lib-musl` instead
+2. Patch the existing port of mimalloc v1.6.3 to adapt to the latest Unikraft's memory allocation interface
```
Comment (RaduNichita) [inline content/blog/2024-06-16-unikraft-gsoc-port-mimalloc-to-unikraft.mdx:28]:
```suggestion
1. Patch the existing port of mimalloc v1.6.3 to adapt to the latest Unikraft's memory allocation interface
```

File: content/blog/2024-06-16-unikraft-gsoc-port-mimalloc-to-unikraft.mdx [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+3. Patch the latest mimalloc (v2.1.7) to the latest Unikraft version (v0.17.0)
```
Comment (RaduNichita) [inline content/blog/2024-06-16-unikraft-gsoc-port-mimalloc-to-unikraft.mdx:29]:
```suggestion
1. Patch the latest mimalloc (v2.1.7) to the latest Unikraft version (v0.17.0)
```

File: content/blog/2024-06-16-unikraft-gsoc-port-mimalloc-to-unikraft.mdx [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+Even though the mimalloc library can now be compiled with the Musl library on the latest Unikraft core without errors, its functionality, portability, and performance remain untested. My future work includes:
+
+1. Test the mimalloc memory allocator aggressively, both with and without virtual memory enabled.
+2. Add support for Unikraft's other memory allocation interface that are not implemented yet by the current port (such as `maxalloc()`, `availmem()`, and `addmem()`).
+3. Port the latest version of mimalloc to the latest Unikraft core.
+
+# Acknowledgements
+
+Special thanks to Razvan Virtan and Radu Nichita, my two amazing mentors, for their support along the way. I would also like to thank Razvan Deaconescu, Hugo Lefeuvre, Stefan Jumarea, Marco Schlumpp, Sergiu Moga, and the entire Unikraft community for all the work, discussions, and guidance which made this project possible.
```
Comment (RaduNichita) [inline content/blog/2024-06-16-unikraft-gsoc-port-mimalloc-to-unikraft.mdx:63]:
```suggestion
Special thanks to Răzvan Vîrtan and Radu Nichita, my two amazing mentors, for their support along the way. I would also like to thank Razvan Deaconescu, Hugo Lefeuvre, Ștefan Jumarea, Marco Schlumpp, Sergiu Moga, and the entire Unikraft community for all the work, discussions, and guidance which made this project possible.
```

File: content/blog/2024-06-16-unikraft-gsoc-port-mimalloc-to-unikraft.mdx [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+## Other works
+
+While building and testing my implementation, I also ran into some minor bugs with the `ukvmem` library, addressed in [this PR](https://github.com/unikraft/unikraft/pull/1447).
+
+I also fixed a few typos in the documentation and added notes in the [Unikraft Internals guide](https://unikraft.org/guides/internals) about enabling hardware acceleration with KVM for random number generation since [Unikraft v0.17.0](https://github.com/unikraft/unikraft/releases/tag/RELEASE-0.17.0). Please see [this PR](https://github.com/unikraft/docs/pull/405).
+
+# Next steps
+
+Even though the mimalloc library can now be compiled with the Musl library on the latest Unikraft core without errors, its functionality, portability, and performance remain untested. My future work includes:
```
Comment (RaduNichita) [inline content/blog/2024-06-16-unikraft-gsoc-port-mimalloc-to-unikraft.mdx:55]:
```suggestion
Even though the mimalloc library can now be compiled with the Musl library on the latest Unikraft core without errors, its functionality, portability, and performance have to be tested. My future work includes:
```

File: content/blog/2024-06-16-unikraft-gsoc-port-mimalloc-to-unikraft.mdx [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+1. Test the mimalloc memory allocator aggressively, both with and without virtual memory enabled.
+2. Add support for Unikraft's other memory allocation interface that are not implemented yet by the current port (such as `maxalloc()`, `availmem()`, and `addmem()`).
```
Comment (RaduNichita) [inline content/blog/2024-06-16-unikraft-gsoc-port-mimalloc-to-unikraft.mdx:58]:
```suggestion
1. Add support for Unikraft's other memory allocation interface that are not implemented yet by the current port (such as `maxalloc()`, `availmem()`, and `addmem()`).
```

File: content/blog/2024-06-16-unikraft-gsoc-port-mimalloc-to-unikraft.mdx [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+3. Port the latest version of mimalloc to the latest Unikraft core.
```
Comment (RaduNichita) [inline content/blog/2024-06-16-unikraft-gsoc-port-mimalloc-to-unikraft.mdx:59]:
```suggestion
1. Port the latest version of mimalloc to the latest Unikraft core.
```

File: content/blog/2024-06-16-unikraft-gsoc-port-mimalloc-to-unikraft.mdx [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+After discussing with the community, we decided to add the `stdatomic.h` header to the `musl` library as a solution. This is done by first creating the header file in the `include/` directory in the `lib-musl` port, and then setting `LIBMUSL_GLOBAL_INCLUDES-y += -I$(LIBMUSL_BASE)/include` in the lib-musl `Makefile.uk`. Please see [this draft PR](https://github.com/unikraft/lib-musl/pull/80) for details of the work, and [this blog](https://unikraft.org/docs/internals/build-system#makefileuk) for more on how Unikraft's `Makefile.uk` works.
+
+## Adapting `lib-mimalloc` to `lib-musl` and Unikraft's new threading interface
+
+This work includes changing the dependency configurations of the port, fixing a few variable names, and including a new header to prevent compilation errors. All well-illustrated in [this draft PR](https://github.com/unikraft/lib-mimalloc/pull/12).
+
+## Other works
```
Comment (RaduNichita) [inline content/blog/2024-06-16-unikraft-gsoc-port-mimalloc-to-unikraft.mdx:47]:
```suggestion
## Other work
```

**Review by RaduNichita — COMMENTED** [review]
Nice job, @huyang531. 

I did a first pass on the blogpost and added a few comments.

Please make sure to check all the markdown errors reported by the Github Action.

Could you also add a few more phrases about the benefit of the community if we have more memory allocators,

**Comment by huyang531** [comment]
All done! @RaduNichita

**Comment by StefanJum** [comment]
@huyang531 you need to quash all the commits. The PR should only have one commit in the end, with all the changes.

@RaduNichita @razvanvirtan please take another look content-wise, so we can merge and publish today.

**Comment by StefanJum** [comment]
@huyang531 please add a proper name and message to the commit.
Should be something like `blog: Add post on porting mimalloc to Unikraft`.

**Review by StefanJum — APPROVED** [review]

**Pull request merged**